File: dao/redis/ratelimit.go
package redis

import (
	"crypto/rand"
	_ "embed"
	"encoding/hex"
	"fmt"
	"strconv"
	"time"

	"github.com/go-redis/redis"
)

//go:embed ratelimit.lua
var voteRateLimitLua string

var voteRateLimitScript = redis.NewScript(voteRateLimitLua)

// VoteRateLimitResult 描述一次限流判定的结果。
type VoteRateLimitResult struct {
	Allowed    bool
	Count      int64
	RetryAfter time.Duration
}

// AllowVoteRequest 用「精确滑动窗口」判定某个用户在当前窗口内是否还有投票额度。
//
// 之所以用 ZSET 记录每次请求的时间戳，而不是固定窗口计数：
// 固定窗口在两个窗口的交界处会放过接近 2 倍限额的突发流量，
// 滑动窗口只统计最近 window 内的请求，边界更严格。
func AllowVoteRequest(userID string, window time.Duration, limit int64) (*VoteRateLimitResult, error) {
	if limit <= 0 {
		return nil, fmt.Errorf("vote rate limit must be positive, got %d", limit)
	}
	now := time.Now().UnixMilli()
	// A process-local sequence is not unique across multiple application
	// instances. A random nonce keeps ZSET members distinct even when two
	// instances handle requests for the same user in the same millisecond.
	var nonce [16]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return nil, fmt.Errorf("generate vote rate-limit member: %w", err)
	}
	member := strconv.FormatInt(now, 10) + "-" + hex.EncodeToString(nonce[:])

	raw, err := voteRateLimitScript.Run(client,
		[]string{getRedisKey(KeyVoteRateLimitPrefix + userID)},
		now, window.Milliseconds(), limit, member,
	).Result()
	if err != nil {
		return nil, err
	}
	values, ok := raw.([]interface{})
	if !ok || len(values) < 3 {
		return nil, fmt.Errorf("unexpected vote rate limit script result: %#v", raw)
	}
	allowed, err := redisInt64(values[0])
	if err != nil {
		return nil, err
	}
	count, err := redisInt64(values[1])
	if err != nil {
		return nil, err
	}
	retryAfterMs, err := redisInt64(values[2])
	if err != nil {
		return nil, err
	}
	return &VoteRateLimitResult{
		Allowed:    allowed == 1,
		Count:      count,
		RetryAfter: time.Duration(retryAfterMs) * time.Millisecond,
	}, nil
}

File: dao/redis/ratelimit.lua
-- 用户级滑动窗口限流（精确型：ZSET 日志）
-- 返回 {是否放行, 窗口内当前请求数, 建议重试等待毫秒}
--
-- KEYS[1] 限流 key，按用户维度隔离
-- ARGV[1] 当前时间戳（毫秒）
-- ARGV[2] 窗口长度（毫秒）
-- ARGV[3] 窗口内允许的最大请求数
-- ARGV[4] 本次请求的唯一成员（毫秒时间戳-随机 nonce）
local key = KEYS[1]
local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local member = ARGV[4]

-- 1. 把滑出窗口的历史请求清掉，只保留 [now-window, now]
redis.call('ZREMRANGEBYSCORE', key, 0, now - window)

-- 2. 统计窗口内剩余请求数
local count = redis.call('ZCARD', key)

-- 3. 超限：不写入，返回最早一次请求的剩余等待时间
if count >= limit then
    local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
    local retry_after = window
    if oldest[2] then
        retry_after = tonumber(oldest[2]) + window - now
        if retry_after < 0 then
            retry_after = 0
        end
    end
    return {0, count, retry_after}
end

-- 4. 放行：记录本次请求并把整个 key 续期到窗口之后
redis.call('ZADD', key, now, member)
redis.call('PEXPIRE', key, window)
return {1, count + 1, 0}
